Compact policy key

jrajahalme opened the pull request:
Make MapState `Key` type smaller so that it fits into 64 bits, and only use the non-Identity part (32 bits) in `bitlpm.Trie`. These changes make a mapstate benchmark >10% faster due to more efficient use of memory.

While we are at it, change the Key types like so:
- direction: `uint8` -> `trafficdirection.TrafficDirection`
- identity: `uint32` -> `identity.NumericIdentity`
- protocol: `uint8` -> `u8proto.U8proto`

This change adds type safety and eliminates a lot of type conversions throughout the code base.

Finally, add builder-style helpers for creating `Key`s.

Note: Please review by commit, as the first commits contain quite a bit of refactoring, the later commits contain the more interesting changes that are easier to review in isolation.

jrajahalme committed:
policy: Hide more internals

Make Key.trafficDirection private and access it with
Key.TrafficDirection(). This allows squeezing traffic direction into a
single bit in a later commit.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Compact Key

Make Key smaller so that less memory is used for them. This is done by
storing port prefix length instead of the inverted mask, and keeping the
traffic direction bit in the highest bit of the same byte ('bits').

Large benchmark test (BenchmarkRegenerateCIDRDenyPolicyRules) uses 4%
less memory, but is ~7% faster with this change.

policymap prototypes are reorganized so that the parameters appear in the
"network stack order":

- traffic direction
- L3 identity
- L4 protocol
- L4 port
- port prefix length
+ non-key values

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Split lpm part of the key into LPMKey

bitlpm.Trie stores the Keys in nodes, so it is beneficial to not keep the
always zeroed Indentity in the key held in the trie. Now that Keys are
only ever constructed with types.NewKey, it is easy to split the
longest-prefix-match parts into LPMKey.

This change alone makes mapstate benchamrks upto 5% faster.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Hide key internals

Make Key.InvertedPortMask private with utility functions for creating
Keys.

Also define a type for the simple Go map for MapState.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

squeed commented on pkg/policy/mapstate.go:
```diff
@@ -32,24 +32,42 @@ import (
+type MapStateMap map[Key]MapStateEntry
+
+// NewKey returns an ingress key for the given parameters
+func NewKey(direction uint8, identity uint32, proto uint8, port uint16, prefixLen uint8) Key {
```
Suggestion: could we use a builder interface? something like

`IngressKey().TCPPort(80).Identity(42)`

jrajahalme replied:
Would `IngressKey().WithIdentity(42)` be clearer?

jrajahalme replied:
Added a new commit with this change to the end.

squeed commented on pkg/policy/types/types.go:
```diff
@@ -100,86 +87,61 @@ func (k Key) IsEgress() bool {
+		return 1 // direction always specified
 	}
-	return keyPrefix
+	return 9 + uint(k.PortPrefixLen())
```
Can you comment what this number is for?

jrajahalme replied:
Added comment:
```
	// 1 bit from direction bit + 8 bits for the protocol when `k.Nexthdr' != 0
```

kaworu reviewed (approved):
Hubble test cleanup LGTM

derailed commented on pkg/policy/types/types.go:
```diff
@@ -17,164 +19,226 @@ const MapStatePrefixLen = uint(32)
+}
+
+func (k Key) WithTCPPort(port uint16) Key {
+	k = k.WithProto(u8proto.TCP)
```
Perhaps we could dry these up ie something like withPortProto(u8proto.U8proto, uint16)?

jrajahalme replied:
Good point, no point repeating any functionality here!

jrajahalme replied:
These have to be exported, though. Used for testing from multiple packages.

derailed reviewed (approved):
@jrajahalme Nice work!

jrajahalme committed:
policy: Change Identity and Nexthdr types

Key Identity and Nexthdr fields are now defined as
identity.NumericIdentity and u8proto.U8proto, respectively. This improves
type safety and almost eliminates the need for type casts regarding
these.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Builders for policy Keys

Provide builder-style utility functions for composing policy keys.

Old style key utilities are still used in policy package internal tests.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

mhofstetter reviewed (approved):
`cilium/proxy` & `cilium/envoy` :heavy_check_mark:  Thanks!
